File: src/main/java/org/apache/airavata/credentials/model/SSHUserCredential.java
package org.apache.airavata.credentials.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.ForeignKey;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;

@Entity
public class SSHUserCredential {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String sshCredentialId;

    @Column(nullable = false)
    private String username;

    // Many-to-one rather than one-to-one: the same registered key can back
    // credentials
    // for different usernames, and deleting a credential must not take the key with
    // it.
    @ManyToOne
    @JoinColumn(name = "ssh_key_id", foreignKey = @ForeignKey(name = "fk_credential_ssh_key"))
    private SSHKeyEntity sshKey;

    public String getSshCredentialId() {
        return sshCredentialId;
    }

    public void setSshCredentialId(String sshCredentialId) {
        this.sshCredentialId = sshCredentialId;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public SSHKeyEntity getSshKey() {
        return sshKey;
    }

    public void setSshKey(SSHKeyEntity sshKey) {
        this.sshKey = sshKey;
    }
}

File: api/credentials/model/ssh_user_credential.go
package model

import (
	"github.com/google/uuid"
	"gorm.io/gorm"
)

// SSHUserCredential binds a username to a registered SSH key.
//
// The link to SSHKey is many-to-one rather than one-to-one on purpose: the same key
// can back credentials for different usernames, and deleting a credential must not
// take the key with it. The reverse protection — refusing to delete a key that is
// still referenced — is enforced in the service layer as a 409, not by the schema.
//
// Java: org.apache.airavata.credentials.model.SSHUserCredential
type SSHUserCredential struct {
	ID string `gorm:"column:ssh_credential_id;primaryKey;type:varchar(36)" json:"sshCredentialId"`

	Username string `gorm:"column:username;type:varchar(255);not null" json:"username"`

	SSHKeyID *string `gorm:"column:ssh_key_id;type:varchar(36);index" json:"sshKeyId,omitempty"`
	SSHKey   *SSHKey `gorm:"references:ID;constraint:OnDelete:RESTRICT,OnUpdate:CASCADE" json:"sshKey,omitempty"`
}

// TableName returns the table backing SSHUserCredential.
func (SSHUserCredential) TableName() string { return "ssh_user_credentials" }

// BeforeCreate assigns a UUID when none was supplied.
func (c *SSHUserCredential) BeforeCreate(*gorm.DB) error {
	if c.ID == "" {
		c.ID = uuid.NewString()
	}
	return nil
}
